### PR TITLE
Implement tests

### PR DESCRIPTION
**Pull request by felipebergamin**
[Work in Progress]

This pull request adds testing libraries and implement tests with 100% coverage.

To do list:

- [x] Install @testing-library/react-hooks
- [x] Write tests for useTTL
- [x] Write tests for useQueryWithTTL
- [x] Write tests for useLazyQueryWithTTL

### CODE REVIEW

File: .github/workflows/testing.js.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+on:
+  pull_request:
+    branches: [ master ]
+  push:
+    branches: [ master ]
+
+jobs:
+  build:
+
+    runs-on: ubuntu-20.04
```
Comment (cabelitos) [inline .github/workflows/testing.js.yml:12]: not sure if possible, but maybe we can use alpine in here? It's much smaller than ubuntu. maybe it can save some download time.

File: __tests__/helpers.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// eslint-disable-next-line import/prefer-default-export
+export const wait = async (ms: number): Promise<void> =>
```
Comment (cabelitos) [inline __tests__/helpers.ts]: please, use export default (avoid disabling lint rules)

File: src/use-query-with-ttl.ts [hunk 3]
```diff
@@ -37,7 +37,7 @@ const useQueryWithTTL: HookWithTTLFunction = ({
     fetchPolicy,
     onCompleted: data => {
       refreshTtl();
-      if (queryOptions?.onCompleted) {
+      if (queryOptions.onCompleted) {
```
Comment (cabelitos) [inline src/use-query-with-ttl.ts:40]: why this change?

File: src/use-lazy-query-with-ttl.ts [hunk 4]
```diff
@@ -16,7 +16,7 @@ const useLazyQueryWithTTL: HookWithTTLFunction<'lazy'> = ({
     fetchPolicy,
     onCompleted: data => {
       refreshTtl();
-      if (queryOptions?.onCompleted) {
+      if (queryOptions.onCompleted) {
```
Comment (cabelitos) [inline src/use-lazy-query-with-ttl.ts:19]: ditto

File: __tests__/use-ttl.test.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import { renderHook } from '@testing-library/react-hooks';
+
+import { wait } from './helpers';
+import useTTL from '../src/use-ttl';
+
+describe('useTTL hook', () => {
+  it('should properly handle TTL status - expired/non expired', async () => {
+    const { result, rerender } = renderHook(() => useTTL(100));
+    expect(result.current[0]).toBe(false);
+    await wait(110);
```
Comment (cabelitos) [inline __tests__/use-ttl.test.ts]: instead of using a real timeout can't you mock the clock itself? https://sinonjs.org/releases/v1.17.6/fake-timers/

File: __tests__/use-lazy-query-with-ttl.test.tsx [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  it('should ignore cache and fetch data from server when TTL expires', async () => {
+    const { getByText, rerender, getByTestId } = render(
+      <MockedProvider cache={cache} mocks={mock}>
+        <LazyQueryComponent />
+      </MockedProvider>,
+    );
+    fireEvent.click(getByTestId('fetch-button'));
+    await waitFor(() => expect(getByText(/Todo from cache/)).toBeTruthy());
+
+    await wait(200);
```
Comment (cabelitos) [inline __tests__/use-lazy-query-with-ttl.test.tsx]: IMO this should be Constants.hookParams.ttl (import the constant and use it) + some time.
By using the constant, we make sure that this will never fail due to someone changing the hook constant and forgetting to change the tests.

However I highly recommend that you use sinon fake timers.

File: __tests__/use-lazy-query-with-ttl.test.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
+/* eslint-disable import/no-extraneous-dependencies */
```
Comment (cabelitos) [inline __tests__/use-lazy-query-with-ttl.test.tsx]: add eslint rule for extraneuous deps. Put a whitelist for the testing stuff.
Avoid disabling eslint rules.

File: __tests__/stubs/NormalQueryComponent.tsx [hunk 8]
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+
+// eslint-disable-next-line no-restricted-imports
```
Comment (cabelitos) [inline __tests__/stubs/NormalQueryComponent.tsx]: avoiding disabling rules. Change the rule to allow ../../ or use typescript path aliases.

File: .github/workflows/testing.js.yml (hunk 1 above)
Comment (felipebergamin) [reply .github/workflows/testing.js.yml:12]: I think alpine isn't supported

https://docs.github.com/pt/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources

File: __tests__/use-ttl.test.ts (hunk 5 above)
Comment (felipebergamin) [reply __tests__/use-ttl.test.ts]: done

File: src/use-query-with-ttl.ts (hunk 3 above)
Comment (felipebergamin) [reply src/use-query-with-ttl.ts:40]: it was an unnecessary optional chaining that was creating a branch which could not be covered by tests, because query options has a default value of `{}` and will never be `undefined`

File: src/use-lazy-query-with-ttl.ts (hunk 4 above)
Comment (felipebergamin) [reply src/use-lazy-query-with-ttl.ts:19]: this optional chaining was creating a branch which could not be covered by tests, because query options has a default value of `{}` and will never be `undefined`

**Review by cabelitos — APPROVED** [review]
